frame/contracts/src/wasm/mod.rs

```diff
@@ -122,8 +128,25 @@ where
 		original_code: Vec<u8>,
 		schedule: &Schedule<T>,
 		owner: AccountIdOf<T>,
-	) -> Result<Self, &'static str> {
-		prepare::prepare_contract(original_code, schedule, owner)
+	) -> Result<Self, (DispatchError, &'static str)> {
+		let module =
+			prepare::prepare_contract(original_code, schedule, owner).map_err(|(err, msg)| {
+				if !msg.is_empty() {
+					log::debug!(
+						target: "runtime::contracts",
+						"CodeRejected: {}",
+						msg,
+					);
+				}
+				(err, msg)
+			})?;
+		// When instrumenting a new code we apply a stricter limit than enforced by the
+		// `RelaxedCodeVec` in order to leave some headroom for reinstrumentation.
+		ensure!(
+			module.code.len() as u32 <= T::MaxCodeLen::get(),
+			(<Error<T>>::CodeTooLarge.into(), ""),
+		);
```

Comment:
I tried to describe it in the comment above this: When instrumenting a new code we apply a stricter limit than what is enforced by the `RelaxedCodeVec`.

Reply:
Yes I understood the comment, I'm just saying that AFAICT it would already fail [here](https://github.com/paritytech/substrate/blob/395149c20c73a44c80c83f7481b16a3f93734e97/frame/contracts/src/wasm/prepare.rs#L409) if the code is larger than `MaxCodeLen`

Reply:
but after that the `code` gets instrumented, and probably could get and increased size, which is being bounded on [this](https://github.com/paritytech/substrate/blob/395149c20c73a44c80c83f7481b16a3f93734e97/frame/contracts/src/wasm/prepare.rs#L418) line only to `RelaxedCodeVec` max. size

Reply:
@agryaznov exactly. The check @ascjones is against the pristine code. It gets bigger after instrumentation. This is the whole point of the `RelaxedCodeVec`. To leave some headroom in case a automatic re-instrumentation happens. If that would make the contract too big it would brick.